transport/server.cc

```diff
@@ -724,11 +724,36 @@ future<> cql_server::connection::process_request() {
         }
 
         auto fut = get_units(_server._memory_available, mem_estimate);
+        struct request_releaser {
+            timer<lowres_clock> timer;
+            bool timed_out = false;
+        };
+        ::shared_ptr<request_releaser> releaser;
         if (_server._memory_available.waiters()) {
+            releaser = ::make_shared<request_releaser>();
+            releaser->timer.set_callback([this, releaser, mem_estimate] {
+                releaser->timed_out = true;
+                _server._memory_available.signal(mem_estimate);
+                ++_server._stats.requests_shed;
+            });
+            //FIXME: Instead of a hardcoded value, it would be nice to parse the timeout from the message - which may require CQL parsing, i.e. not trivial.
+            //FIXME: Concurrency of this path is 1 and thus releasing 1 request per 5 seconds makes very little sense on its own. It would be better if requests already
+            // had an educated guess how long they stayed in the queue - e.g. by the upper layer buffering N requests from the socket, or by using SO_TIMESTAMP information if available.
+            releaser->timer.arm(lowres_clock::now() + std::chrono::seconds(5));
             ++_server._stats.requests_blocked_memory;
         }
 
-        return fut.then([this, length = f.length, flags = f.flags, op, stream, tracing_requested] (semaphore_units<> mem_permit) {
+        return fut.then([this, length = f.length, flags = f.flags, op, stream, tracing_requested, releaser = std::move(releaser)] (semaphore_units<> mem_permit) {
+          if (releaser) {
+              if (releaser->timed_out) {
+                // Semaphore units were already returned in the timeout handler, so it's important to release the units here.
+                mem_permit.release();
```

Comment:
Make mem_permit part of request_releaser, and it will happen by itself. But IMO you are doing too many things behind semaphore's back.

Why not use semaphore::wait(duration)?

Reply:
Sigh, didn't know we have a timed wait in our semaphore. It's obviously a much better choice.